File: albedo-math/src/main/java/jp/albedo/jpl/kernel/tree/TreeNode.java
package jp.albedo.jpl.kernel.tree;

import java.util.Collections;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.Map;
import java.util.Optional;

public class TreeNode<NodeType, EdgeType> {

    private final NodeType nodeValue;

    private EdgeType incomingEdgeValue;

    private final Map<NodeType, TreeNode<NodeType, EdgeType>> children = new HashMap<>();

    public TreeNode(NodeType nodeValue, EdgeType incomingEdgeValue) {
        this.nodeValue = nodeValue;
        this.incomingEdgeValue = incomingEdgeValue;
    }

    public NodeType getNodeValue() {
        return nodeValue;
    }

    public void setIncomingEdgeValue(EdgeType incomingEdgeValue) {
        if (this.incomingEdgeValue != null) {
            throw new IllegalStateException("Node already has edge value!");
        }
        this.incomingEdgeValue = incomingEdgeValue;
    }

    public EdgeType getIncomingEdgeValue() {
        return incomingEdgeValue;
    }

    public void append(NodeType nodeValue, TreeNode<NodeType, EdgeType> child) {
        children.put(nodeValue, child);
    }

    public boolean hasChild(NodeType nodeValue) {
        return children.containsKey(nodeValue);
    }

    public Map<NodeType, TreeNode<NodeType, EdgeType>> getChildren() {
        return Collections.unmodifiableMap(children);
    }

    public Optional<TreeNode<NodeType, EdgeType>> getChild(NodeType child) {
        return Optional.ofNullable(children.get(child));
    }

    public Optional<TreeNode<NodeType, EdgeType>> findNode(NodeType toFind) {
        if (children.containsKey(toFind)) {
            return Optional.of(children.get(toFind));
        }
        return children.values().stream()
                .map(childNode -> childNode.findNode(toFind))
                .filter(Optional::isPresent)
                .map(Optional::get)
                .findFirst();
    }

    public Optional<LinkedList<TreeNode<NodeType, EdgeType>>> findBranch(NodeType toFind) {
        if (nodeValue.equals(toFind)) {
            return Optional.of(new LinkedList<>(Collections.singletonList(this)));
        }

        return children.values().stream()
                .map(childNode -> childNode.findBranch(toFind))
                .filter(Optional::isPresent)
                .map(Optional::get)
                .peek(branch -> branch.addFirst(this))
                .findFirst();
    }

}

File: packages/core/src/jpl/kernel/tree/TreeNode.ts
export class TreeNode<NodeType, EdgeType> {

    readonly value: NodeType;
    incomingEdge?: EdgeType;
    readonly children: Map<NodeType, TreeNode<NodeType, EdgeType>> = new Map();

    constructor(value: NodeType, incomingEdge?: EdgeType) {
        this.value = value;
        this.incomingEdge = incomingEdge;
    }

    setIncomingEdge(incomingEdge: EdgeType): void {
        if (this.incomingEdge != null) {
            throw Error('Node already has edge value!');
        }
        this.incomingEdge = incomingEdge;
    }

    append(nodeValue: NodeType, child: TreeNode<NodeType, EdgeType>): void {
        this.children.set(nodeValue, child);
    }

    hasChild(nodeValue: NodeType): boolean {
        return this.children.has(nodeValue);
    }

    getChild(nodeValue: NodeType): TreeNode<NodeType, EdgeType> | undefined {
        return this.children.get(nodeValue);
    }

    findNode(toFind: NodeType): TreeNode<NodeType, EdgeType> | undefined {
        if (this.children.has(toFind)) {
            return this.children.get(toFind);
        }

        return Array.from(this.children.values())
            .map((childNode) => childNode.findNode(toFind))
            .find((childNode) => childNode !== undefined);
    }

    findBranchTo(node: NodeType): TreeNode<NodeType, EdgeType>[] | undefined {
        if (this.value === node) {
            return [this];
        }

        const branch = Array.from(this.children.values())
            .map((childNode) => childNode.findBranchTo(node))
            .find((branch) => branch !== undefined);

        branch?.unshift(this);

        return branch;
    }

};
